extensions-contrib/gce-extensions/src/main/java/org/apache/druid/indexing/overlord/autoscaling/gce/GCEAutoScaler.java

```diff
@@ -0,0 +1,496 @@
+
+      Compute computeService = createComputeService();
+      Compute.InstanceGroupManagers.ListManagedInstances request =
+              computeService
+                      .instanceGroupManagers()
+                      .listManagedInstances(project, zone, managedInstanceGroupName);
+      // Notice that while the doc says otherwise, there is not nextPageToken to page
+      // through results and so everything needs to be in the same page
+      request.setMaxResults(500L); // 500 is sadly the max
+      InstanceGroupManagersListManagedInstancesResponse response = request.execute();
+      for (ManagedInstance mi : response.getManagedInstances()) {
+        ids.add(GCEUtils.extractNameFromInstance(mi.getInstance()));
+      }
+      log.debug("Found running instances [%s]", String.join(",", ids));
+    }
+    catch (Exception e) {
+      log.error(e, "Unable to get instances.");
+    }
+    return ids;
+  }
+
+  /**
+   * Converts the IPs to IDs
+   */
+  @Override
+  public List<String> ipToIdLookup(List<String> ips)
+  {
+    log.debug("Asked IPs -> IDs for: [%s]", String.join(",", ips));
+
+    if (ips.isEmpty()) {
+      return new ArrayList<>();
+    }
+
+    // If the first one is not an IP, just assume all the other ones are not as well and just
+    // return them as they are. This check is here because Druid does not check if IPs are
+    // actually IPs and can send IDs to this function instead
+    if (!InetAddresses.isInetAddress(ips.get(0))) {
+      log.debug("Not IPs, doing nothing");
+      return ips;
+    }
+
+    final String project = envConfig.getProjectId();
+    final String zone = envConfig.getZoneName();
+    try {
+      Compute computeService = createComputeService();
+      Compute.Instances.List request = computeService.instances().list(project, zone);
+      // Cannot filter by IP atm, see below
+      // request.setFilter(GCEUtils.buildFilter(ips, "networkInterfaces[0].networkIP"));
+
+      List<String> instanceIds = new ArrayList<>();
+      InstanceList response;
+      do {
+        response = request.execute();
+        if (response.getItems() == null) {
+          continue;
+        }
+        for (Instance instance : response.getItems()) {
+          // This stupid look up is needed because atm it is not possible to filter
+          // by IP, see https://issuetracker.google.com/issues/73455339
+          if (ips.contains(instance.getNetworkInterfaces().get(0).getNetworkIP())) {
+            instanceIds.add(instance.getName());
+          }
+        }
+        request.setPageToken(response.getNextPageToken());
+      } while (response.getNextPageToken() != null);
+
+      log.debug("Converted to [%s]", String.join(",", instanceIds));
+      return instanceIds;
+    }
+    catch (Exception e) {
+      log.error(e, "Unable to convert IPs to IDs.");
+    }
+
+    return new ArrayList<>();
+  }
+
+  /**
+   * Converts the IDs to IPs - this is actually never called from the outside but it is called once
```

Comment:
why it is public?

Reply:
it is part of the interface we inherit from